spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/ManagedCommitSuite.scala

```diff
@@ -116,6 +123,405 @@ class ManagedCommitSuite
+  // 4. Write commit 3/4 using new commit owner.
+  // 5. Read the table again and make sure right APIs are called:
+  //    a) If read query is run in scala, we do listing 2 times. So CS2.getCommits will be called
+  //       twice. We should not be contacting CS1 anymore.
+  //    b) If read query is run on SQL, we do listing only once. So CS2.getCommits will be called
+  //       only once.
+  test("snapshot is updated recursively when owner changes multiple times") {
+    val batchSize = 10
+    val underlyingCommitStore2 = new InMemoryCommitStore(batchSize)
+    val underlyingCommitStore1 = new InMemoryCommitStore(batchSize) {
+      override def commitImpl(
+          logStore: LogStore,
+          hadoopConf: Configuration,
+          logPath: Path,
+          commitVersion: Long,
+          commitFile: FileStatus,
+          commitTimestamp: Long): CommitResponse = {
+        super.commitImpl(logStore, hadoopConf, logPath, commitVersion, commitFile, commitTimestamp)
+        underlyingCommitStore2
+          .commitImpl(logStore, hadoopConf, logPath, commitVersion, commitFile, commitTimestamp)
+      }
+
+      override protected[delta] def registerBackfill(
+          logPath: Path, backfilledVersion: Long): Unit = {
+        super.registerBackfill(logPath, backfilledVersion)
+        underlyingCommitStore2.registerBackfill(logPath, backfilledVersion)
+      }
+    }
+    val cs1 = new TrackingCommitStore(underlyingCommitStore1)
+    val cs2 = new TrackingCommitStore(underlyingCommitStore2)
+
+    case class TrackingInMemoryCommitStoreBuilder(
+        override val name: String,
+        commitStore: CommitStore) extends CommitStoreBuilder {
+      var numBuildCalled = 0
+      override def build(conf: Map[String, String]): CommitStore = {
+        numBuildCalled += 1
+        commitStore
+      }
+    }
+    val builder1 = TrackingInMemoryCommitStoreBuilder(name = "tracking-in-memory-1", cs1)
+    val builder2 = TrackingInMemoryCommitStoreBuilder(name = "tracking-in-memory-2", cs2)
+    Seq(builder1, builder2).foreach(CommitStoreProvider.registerBuilder)
+
+    def resetMetrics(): Unit = {
+      Seq(builder1, builder2).foreach { b => b.numBuildCalled = 0 }
+      Seq(cs1, cs2).foreach(_.reset())
+    }
+
+    withSQLConf(
+        MANAGED_COMMIT_OWNER_NAME.defaultTablePropertyKey -> builder1.name) {
+      withTempDir { tempDir =>
+        val tablePath = tempDir.getAbsolutePath
+        // Step-1: Make 3 commits on the table with CS1 as owner.
+        Seq(0).toDF.write.format("delta").mode("append").save(tablePath) // version 0
+        Seq(1).toDF.write.format("delta").mode("append").save(tablePath) // version 1
+        Seq(2).toDF.write.format("delta").mode("append").save(tablePath) // version 2
+        DeltaLog.clearCache()
+        checkAnswer(sql(s"SELECT * FROM delta.`$tablePath`"), Seq(Row(0), Row(1), Row(2)))
+        def deltaLog(): DeltaLog = DeltaLog.forTable(spark, tablePath)
```

Comment:
I only see one call to this method (L350 below) and also see code reusing `log`. Maybe this method isn't needed?